In [1]:
from keras.preprocessing.image import ImageDataGenerator
from keras.models import Sequential
from keras.layers.normalization import BatchNormalization
from keras.layers.convolutional import Conv2D
from keras.layers.convolutional import MaxPooling2D
from keras.layers.core import Activation
from keras.layers.core import Flatten
from keras.layers.core import Dropout
from keras.layers.core import Dense
from keras.utils import np_utils
import tensorflow as tf

C:\Users\Balajisri\Anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


# 1. Read the images and generate the train and test dataset
# 2. Divide the data set into Train and validation data sets

In [22]:
image_path = r'D:\Development\Plant Seed\train'

train_datagen = ImageDataGenerator(rescale = 1./255,
                                   shear_range = 0.2,
                                   zoom_range = 0.2, validation_split=0.1,
                                   horizontal_flip = True)

training_set = train_datagen.flow_from_directory(image_path,
                                                 target_size = (64, 64), 
                                                 batch_size = 15, 
                                                 subset = 'training',
                                                 class_mode = 'categorical')

test_set = train_datagen.flow_from_directory(image_path,
                                            target_size = (64, 64),
                                            batch_size = 15,
                                            subset = 'validation',
                                            class_mode = 'categorical')


Found 4279 images belonging to 12 classes.
Found 471 images belonging to 12 classes.


# Initialize & build the model 

In [23]:
# Initialising the CNN
classifier = Sequential()

# Step 1 - Convolution
classifier.add(Conv2D(32, (3, 3), input_shape = (64, 64, 3), activation = 'relu'))

# Step 2 - Pooling
classifier.add(MaxPooling2D(pool_size = (2, 2)))

#classifier.add(Conv2D(64, (3, 3), activation = 'relu'))
#classifier.add(MaxPooling2D(pool_size = (2, 2)))

# Adding a second convolutional layer
classifier.add(Conv2D(32, (3, 3), activation = 'relu'))
classifier.add(MaxPooling2D(pool_size = (2, 2)))

# Step 3 - Flattening
classifier.add(Flatten())
 
# Step 4 - Full connection
classifier.add(Dense(units = 128, activation = 'relu'))
classifier.add(Dense(units = 12, activation = 'sigmoid'))

# Optimize the model 

In [24]:
classifier.compile(optimizer = 'adam', loss = 'categorical_crossentropy', metrics = ['accuracy'])

# Predict the accuracy for both train and validation data

In [25]:
classifier.fit_generator(training_set, steps_per_epoch=4279,
                         epochs = 3, validation_data=test_set, validation_steps=471)

Epoch 1/3
4279/4279 [==============================] - 576s 135ms/step - loss: 0.9315 - acc: 0.6823 - val_loss: 0.7029 - val_acc: 0.7776
Epoch 2/3
4279/4279 [==============================] - 590s 138ms/step - loss: 0.3490 - acc: 0.8799 - val_loss: 0.6256 - val_acc: 0.8156
Epoch 3/3
4279/4279 [==============================] - 596s 139ms/step - loss: 0.1883 - acc: 0.9340 - val_loss: 0.7312 - val_acc: 0.8248


I have started with initail model with one convolution layer and max pooling with dense which gave me around 70% accuracy. later I have added multiple layer and finally got train accuracy of 93% and validation accuracy of 82%.